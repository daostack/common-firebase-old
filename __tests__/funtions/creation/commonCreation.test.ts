import supertest from 'supertest';
import * as ethers from 'ethers';

import '@functions';
import { env } from '@env';
import { runTest } from '@helpers/runTest';
import { abi } from '@helpers/abi';

runTest((funcs, test) => {
  const create = supertest(funcs.create);
  const { auth } = test;

  const zeroAddress = `0x${'0'.repeat(40)}`;

  const gnosisSafeAddress = '0xEAe13cD0b3586C09AFe762F933dA190D614628b6';
  const createAndAddModulesAddress = '0x51cf75287148cba21FBAAF69FBE3D60546e7d16e';

  const account = '0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1';
  const privateKey = '0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d';

  const idToken = 'eyJhbGciOiJSUzI1NiIsImtpZCI6IjFlNjYzOGY4NDlkODVhNWVkMGQ1M2NkNDI1MzE0Y2Q1MGYwYjY1YWUiLCJ0eXAiOiJKV1QifQ.eyJuYW1lIjoiQWxleGFuZGVyIEl2YW5vdiIsInBpY3R1cmUiOiJodHRwczovL2xoMy5nb29nbGV1c2VyY29udGVudC5jb20vYS0vQU9oMTRHZ2FCeHJMRE9iLWY1TTFLQ1dtVjZ1MzlJXzhoWlFyM0ZHelN3RU1MWmM9czk2LWMiLCJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vY29tbW9uLXN0YWdpbmctNTA3NDEiLCJhdWQiOiJjb21tb24tc3RhZ2luZy01MDc0MSIsImF1dGhfdGltZSI6MTYwMDg1NDM0NiwidXNlcl9pZCI6Ikg1WmtjS0JYNWVYWE55QmlQYXBoOEVIQ2lheDIiLCJzdWIiOiJINVprY0tCWDVlWFhOeUJpUGFwaDhFSENpYXgyIiwiaWF0IjoxNjAwODYxNTc4LCJleHAiOjE2MDA4NjUxNzgsImVtYWlsIjoiYWxleGFuZGVyMjAwMWl2YW5vdkBnbWFpbC5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwiZmlyZWJhc2UiOnsiaWRlbnRpdGllcyI6eyJnb29nbGUuY29tIjpbIjEwNzk1NTE0Mjg2OTQxMTgyNTUwMSJdLCJlbWFpbCI6WyJhbGV4YW5kZXIyMDAxaXZhbm92QGdtYWlsLmNvbSJdfSwic2lnbl9pbl9wcm92aWRlciI6Imdvb2dsZS5jb20ifX0.Rx6GzrRTiZ-JQjnEsSrLr-iwSsJ6d6uGSDdK8jshpSDRCH55_IXz8d1JINmRaR4fJ8fv_SgjKh1nN-HXKt3lJoJKkSNv2OwO0PqXHYafFwfsT1IywU3O6HUui0DwdjYYsLtlcFBQi2ubObohw9B7gW6vzzyQhFWkS4vWMHby86M48T74v_Oj5PhrmYCIdKtjaH6fkK_h4XX2XkDIW9cd9E2lklYwUNMeG9B2mc_dUfC_Z01QMskMHvArBS-haWbXjqXxlVOASN4xCvS6Yym-Kf4TUu_hzzsTX3jMx_uA75GPeuoC7ZUwKz-REsDpvc0llQ7VJwcZ9qq1o4ADZ30Mqw';
  const data = {
    name: 'Test Common',
    founderAddresses: gnosisSafeAddress,
    tokenDist: [0],
    repDist: [100],
    minFeeToJoin: 100,
    fundingGoal: 1000,
    fundingGoalDeadline: new Date().getTime() + 3000,
    byline: 'Testing is life',
    description: 'To test or not to test',
    courseOfAction: 'Test, test and test',
    rules: [],
    links: []
  };

  describe('Common creation', () => {
    it('should be fail creating common transaction without the id token provided', async () => {
      const res = await create.post('/createCommonTransaction');

      expect(res.ok).toBeFalsy();
      expect(res.status).toBe(500);

      expect(res.body.error).not.toBe(null);
      expect(res.body.error.commonMessage).toMatchSnapshot();
    });

    it('should successfully create common creation transaction with all the needed data present', async () => {
      const res = await create
        .post('/createCommonTransaction')
        .send({
          idToken,
          data,
          user: auth.exampleUserRecord()
        });


      expect(res.status).toBe(200);

      expect(res.body.message).toMatchSnapshot();
    });

    it('should successfully create common creation', async () => {
      const res = await create
        .post('/createCommonTransaction')
        .send({
          idToken,
          data,
          user: auth.exampleUserRecord()
        });

      expect(res.status).toBe(200);

      const { toAddress, safeTxHash, encodedData } = res.body;

      const provider = new ethers.providers.JsonRpcProvider(env.blockchain.jsonRpcProvider);

      // Create a wallet to sign the message with
      const wallet = new ethers.Wallet(privateKey, provider);
      const contract = new ethers.Contract(createAndAddModulesAddress, abi, wallet);

      const messageHash = ethers.utils.id(safeTxHash);
      const messageHashBytes = ethers.utils.arrayify(messageHash)

      // Sign the binary data
      const flatSig = await wallet.signMessage(messageHashBytes);

      const contractInstance = await contract.deployed();
      // Call the verifyHash function
      // const transactionResult = await contract.execTransaction(toAddress, '0', encodedData,  0, 0, 0, 0, zeroAddress, zeroAddress, getSignatureBytes(privateKey, messageHash));
      // const transactionResult = await contractInstance.execTransaction(
      //   toAddress,
      //   0,
      //   encodedData,
      //   0,
      //   0,
      //   0,
      //   0,
      //   zeroAddress,
      //   zeroAddress,
      //   flatSig
      //     .replace(/1b$/, '1f')
      //     .replace(/1c$/, '20')
      // );

      // const createCommonRes = await create
      //   .post('/createCommonTransaction')
      //   .send({
      //     idToken,
      //     encodedData,
      //     flatSig,
      //     toAddress
      //   });

      const res2 = await create
        .post('/createCommon')
        .send({
          idToken,
          toAddress,
          encodedData,
          signedData: flatSig
        });

      // console.log(res2);
      console.log(res2)
      expect(res2.status).toEqual(200)
      // expect(res2).toMatchSnapshot();
    });
  });
});

// const in3Common = require('in3-common');
// const ethUtil = require('ethereumjs-util');
//
// const getSignatureBytes = (pk, msgHash) => {
//   const s = ethUtil.ecsign(in3Common.util.toBuffer(msgHash, 32), in3Common.util.toBuffer(pk, 32))
//
//   return in3Common.util.toHex(s.r) + in3Common.util.toHex(s.s).substr(2) + in3Common.util.toHex(s.v).substr(2)
// }
