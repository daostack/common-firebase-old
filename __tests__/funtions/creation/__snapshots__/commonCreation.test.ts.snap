// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Functions Common creation should be fail creating common transaction without the id token provided 1`] = `"Unable to create a common transaction"`;

exports[`Cloud Functions Common creation should successfully create common creation 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "291",
    "content-type": "application/json; charset=utf-8",
    "date": "Thu, 08 Oct 2020 14:14:45 GMT",
    "etag": "W/\\"123-IL03q18ZYzWO51u+MqSdDfh7bro\\"",
    "vary": "Origin",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": Object {
      "encodedData": "0x6cf00fa4000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000003600000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000240000000000000000000000000000000000000000000000000000000000000028000000000000000000000000000000000000000000000000000000000000002c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000b5465737420436f6d6d6f6e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c4bd3a13f6000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005b85574ed1571d2cec8d61a35d394a4d88556f5600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000eae13cd0b3586c09afe762f933da190d614628b600000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000ae000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000160000000000000000000000000000000000000000000000000000000000000090000000000000000000000000000000000000000000000000000000000000009c00000000000000000000000000000000000000000000000000000000000000a8000000000000000000000000000000000000000000000000000000000000000054a6f696e0000000000000000000000000000000000000000000000000000000046756e64696e6752657175657374000000000000000000000000000000000000536368656d65466163746f7279000000000000000000000000000000000000004469637461746f7200000000000000000000000000000000000000000000000052657075746174696f6e41646d696e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007745368e5560000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e1597b856edbbb92f394e08a76b816f96031a1000000000000000000000000000000000000000000000000000000000000000032000000000000000000000000000000000000000000000000000000000012750000000000000000000000000000000000000000000000000000000000000119400000000000000000000000000000000000000000000000000000000000000e1000000000000000000000000000000000000000000000000000000000000004b000000000000000000000000000000000000000000000000000000000000038400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dff3e43710d39d2ba5dda7a8d959ed22cc905b01000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003e8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001da27edf0985e313f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e1597b856edbbb92f394e08a76b816f96031a100000000000000000000000000000000000000000000000000000000000000003200000000000000000000000000000000000000000000000000000000000189c0000000000000000000000000000000000000000000000000000000000007e9000000000000000000000000000000000000000000000000000000000000000e1000000000000000000000000000000000000000000000000000000000000004b00000000000000000000000000000000000000000000000000000000000015180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000175089079d400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000dff3e43710d39d2ba5dda7a8d959ed22cc905b01985e313f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e1597b856edbbb92f394e08a76b816f96031a100000000000000000000000000000000000000000000000000000000000000003200000000000000000000000000000000000000000000000000000000000189c0000000000000000000000000000000000000000000000000000000000007e9000000000000000000000000000000000000000000000000000000000000000e1000000000000000000000000000000000000000000000000000000000000004b0000000000000000000000000000000000000000000000000000000000001518000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005b85574ed1571d2cec8d61a35d394a4d88556f56485cc9550000000000000000000000000000000000000000000000000000000000000000000000000000000000000000bbb06cd354d7f4e67677f090ecc3f6e5916e24476c28e3490000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001da27edf00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000cf8d68f810cb8e3e228a7f31a61bef0c1700a7d50000000000000000000000000000000000000000000000000000000000000000000000000000000000000005000000000000000000000000000000000000000000000000000000000000028400000000000000000000000000000000000000000000000000000000000002040000000000000000000000000000000000000000000000000000000000000204000000000000000000000000000000000000000000000000000000000000004400000000000000000000000000000000000000000000000000000000000000a40000000000000000000000000000000000000000000000000000000000000005000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f000000000000000000000000000000000000000000000000000000000000001f000000000000000000000000000000000000000000000000000000000000001f00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e516d55566776356266316f4e356b364e767275554a766e36366131417a394548634b67443558574272415033564e000000000000000000000000000000000000",
      "idToken": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjFlNjYzOGY4NDlkODVhNWVkMGQ1M2NkNDI1MzE0Y2Q1MGYwYjY1YWUiLCJ0eXAiOiJKV1QifQ.eyJuYW1lIjoiQWxleGFuZGVyIEl2YW5vdiIsInBpY3R1cmUiOiJodHRwczovL2xoMy5nb29nbGV1c2VyY29udGVudC5jb20vYS0vQU9oMTRHZ2FCeHJMRE9iLWY1TTFLQ1dtVjZ1MzlJXzhoWlFyM0ZHelN3RU1MWmM9czk2LWMiLCJpc3MiOiJodHRwczovL3NlY3VyZXRva2VuLmdvb2dsZS5jb20vY29tbW9uLXN0YWdpbmctNTA3NDEiLCJhdWQiOiJjb21tb24tc3RhZ2luZy01MDc0MSIsImF1dGhfdGltZSI6MTYwMDg1NDM0NiwidXNlcl9pZCI6Ikg1WmtjS0JYNWVYWE55QmlQYXBoOEVIQ2lheDIiLCJzdWIiOiJINVprY0tCWDVlWFhOeUJpUGFwaDhFSENpYXgyIiwiaWF0IjoxNjAwODYxNTc4LCJleHAiOjE2MDA4NjUxNzgsImVtYWlsIjoiYWxleGFuZGVyMjAwMWl2YW5vdkBnbWFpbC5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwiZmlyZWJhc2UiOnsiaWRlbnRpdGllcyI6eyJnb29nbGUuY29tIjpbIjEwNzk1NTE0Mjg2OTQxMTgyNTUwMSJdLCJlbWFpbCI6WyJhbGV4YW5kZXIyMDAxaXZhbm92QGdtYWlsLmNvbSJdfSwic2lnbl9pbl9wcm92aWRlciI6Imdvb2dsZS5jb20ifX0.Rx6GzrRTiZ-JQjnEsSrLr-iwSsJ6d6uGSDdK8jshpSDRCH55_IXz8d1JINmRaR4fJ8fv_SgjKh1nN-HXKt3lJoJKkSNv2OwO0PqXHYafFwfsT1IywU3O6HUui0DwdjYYsLtlcFBQi2ubObohw9B7gW6vzzyQhFWkS4vWMHby86M48T74v_Oj5PhrmYCIdKtjaH6fkK_h4XX2XkDIW9cd9E2lklYwUNMeG9B2mc_dUfC_Z01QMskMHvArBS-haWbXjqXxlVOASN4xCvS6Yym-Kf4TUu_hzzsTX3jMx_uA75GPeuoC7ZUwKz-REsDpvc0llQ7VJwcZ9qq1o4ADZ30Mqw",
      "signedData": "0x4a9d3de64b4073d0488bebfbe2d20e64ac86a334241fba1283a79ac80d93a30c3feda4ed5d4ed1c0b5a9d7c379ace94b714de15c99e5d79e48efc37f6011c9b81b",
      "toAddress": "0x5b85574ed1571d2cec8d61a35d394a4d88556f56",
    },
    "headers": Object {
      "content-type": "application/json",
      "user-agent": "node-superagent/3.8.3",
    },
    "method": "POST",
    "url": "http://127.0.0.1:62645/createCommon",
  },
  "status": 500,
  "text": "{\\"error\\":{\\"commonMessage\\":\\"Unable to create common!\\",\\"errorObject\\":\\"{\\\\n    \\\\\\"errorId\\\\\\": \\\\\\"bcafdabe-d27b-409b-8e01-b599180090e0\\\\\\",\\\\n    \\\\\\"name\\\\\\": \\\\\\"CommonError\\\\\\",\\\\n    \\\\\\"userMessage\\\\\\": \\\\\\"Something bad happened\\\\\\",\\\\n    \\\\\\"statusCode\\\\\\": 500\\\\n}\\",\\"error\\":\\"CommonError: invalidIdToken\\"},\\"query\\":{}}",
}
`;

exports[`Cloud Functions Common creation should successfully create common creation transaction with all the needed data present 1`] = `"Create common transaction successfully!"`;
